### PR TITLE
Structured logging

### PR DESCRIPTION
**Pull request by bbguimaraes**
@stevekuznetsov 

Some examples of the output:

https://github.com/bbguimaraes/origin-ci-tool/tree/structured-logging-ex

### CODE REVIEW

File: oct/ansible/oct/callback_plugins/log_results.py [hunk 1]
```diff
@@ -1,16 +1,44 @@
...
+        structlog.processors.format_exc_info,
+        structlog.processors.UnicodeDecoder(),
+        structlog.processors.JSONRenderer(),
+    ],
+    context_class=dict,
+    logger_factory=structlog.stdlib.LoggerFactory(),
+    wrapper_class=structlog.stdlib.BoundLogger,
+    cache_logger_on_first_use=True)
+handler = logging.FileHandler(
+    join(os.environ.get('ANSIBLE_LOG_ROOT_PATH'), 'log.txt'), 'a')
```
Comment (bbguimaraes) [inline oct/ansible/oct/callback_plugins/log_results.py]: Is this used somewhere outside this tool?  Could we change it to a single file?

File: oct/ansible/oct/callback_plugins/log_results.py [hunk 2]
```diff
@@ -380,24 +350,7 @@ def v2_runner_on_async_failed(self, result):
...
-    def v2_playbook_on_no_hosts_remaining(self):
-        """
-        Implementation of the callback endpoint to be
-        fired when a playbook has no hosts remaining
-        to execute on.
-        """
-        self.write_log('No hosts remaining.')
+        logger.error()
 
         # TODO: unimplemented methods follow - do we need them?
```
Comment (bbguimaraes) [inline oct/ansible/oct/callback_plugins/log_results.py]: After a year without any change, can we answer "no" to this?

File: oct/ansible/oct/callback_plugins/log_results.py [hunk 3]
```diff
@@ -1,16 +1,44 @@
...
 from __future__ import absolute_import, division, print_function, unicode_literals
 
-from codecs import open
-from datetime import datetime
-from os import environ, makedirs
-from os.path import exists, join, sep
-from shutil import rmtree
-from tempfile import mkdtemp
-from traceback import format_exc
+import contextlib
```
Comment (stevekuznetsov) [inline oct/ansible/oct/callback_plugins/log_results.py]: In the project we always prefer `from X import Y` style if possible

File: oct/ansible/oct/callback_plugins/log_results.py (hunk 1 above)
Comment (stevekuznetsov) [inline oct/ansible/oct/callback_plugins/log_results.py]: Unfortunately callback plugins need to be self-contained.

File: oct/ansible/oct/callback_plugins/log_results.py [hunk 4]
```diff
@@ -54,182 +79,105 @@ def log_exceptions_v2_playbook_on_start(func):
...
+def sanitize_results(data):
+    """
+    Sanitize the results to remove unnecessary fields.
+    Removes Ansible internal fields that the user doesn't need in their output
+    files and `stdout_lines` if `stdout` is present -- we can let people load
+    the data and format it however they want, but we don't want to store it
+    twice.
+    """
+    for key in data.keys():
+        if key.startswith('_ansible'):
```
Comment (stevekuznetsov) [inline oct/ansible/oct/callback_plugins/log_results.py]: Looks like we may need to make this a recursive sanitize:

```json
{
  "callback": "v2_runner_on_skipped",
  "host": "localhost",
  "result": {
    "msg": "All items completed",
    "changed": false,
    "results": [
      {
        "skipped": true,
        "_ansible_no_log": false,
        "skip_reason": "Conditional check failed",
        "_ansible_item_result": true,
        "item": "origin_ci_inventory_dir",
        "changed": false
},
```

File: oct/ansible/oct/callback_plugins/log_results.py (hunk 2 above)
Comment (stevekuznetsov) [inline oct/ansible/oct/callback_plugins/log_results.py]: Today let's just put a `logger.info()` in these and make sure we implement every upstream callback.

**Comment by stevekuznetsov** [comment]
[(3) [CI] [oct] Implement the event stream logger for the Origin CI Tool](https://trello.com/c/RuVMEjin/128-3-ci-oct-implement-the-event-stream-logger-for-the-origin-ci-tool)

File: oct/ansible/oct/callback_plugins/log_results.py (hunk 1 above)
Comment (bbguimaraes) [inline oct/ansible/oct/callback_plugins/log_results.py]: With the code as it is on this PR, the log file would be at

    ~/.config/origin-ci-tool/logs/log.txt

Can we change it to a single log file, e.g.

    ~/.config/origin-ci-tool/log.txt

or would it impact other places that look for the logs in this location?

There doesn't seem to be a reference to this path on this repository, but I
don't know about other systems.

**Comment by bbguimaraes** [comment]
@stevekuznetsov: updated the code and the examples.

Unfortunately my response to your first comment is now hidden, but please take
a look at it.  I noticed I need to add code to create the directory (if we
decide to keep it a directory), but I'll wait for your response.

File: oct/ansible/oct/callback_plugins/log_results.py (hunk 1 above)
Comment (stevekuznetsov) [inline oct/ansible/oct/callback_plugins/log_results.py]: Single log file sounds good to me, but I think we still need that  environment variable to point to the containing directory. We set it in `oct/config/ansible_client.py` to communicate to the plugins where to put logs.

File: oct/ansible/oct/callback_plugins/log_results.py [hunk 5]
```diff
@@ -150,13 +150,19 @@ def sanitize_results(data):
...
     twice.
     """
-    for key in data.keys():
-        if key.startswith('_ansible'):
-            del data[key]
-    if 'results' in data:
-        for result in data['results']:
-            if 'stdout' in result and 'stdout_lines' in result:
-                del result['stdout_lines']
+    if isinstance(data, list):
```
Comment (stevekuznetsov) [inline oct/ansible/oct/callback_plugins/log_results.py]: Can we have nested `results` fields? If not, we should have a separate recursive function call for sanitizing just the `_ansible` bits -- a smell here is that the first call to `santiize_results()` will not ever be with a list, so we have two separate cases I think -- top-level where you want to get rid of `stdout_lines` and recursively where you want to get rid of keys

**Comment by bbguimaraes** [comment]
@stevekuznetsov: updated.

There are other parts of the code that use the `logs` directory
(`generate_junit.py`, maybe others), so I kept the log file at the same place
and added`some code to the callback module's `__init__` to create the
directory if needed.

If there are no other comments, I think this is ready.

File: oct/ansible/oct/callback_plugins/log_results.py [hunk 6]
```diff
@@ -55,181 +75,125 @@ def log_exceptions_v2_playbook_on_start(func):
...
-    """
-    CALLBACK_VERSION = 2.0
-    CALLBACK_TYPE = 'notification'
-    CALLBACK_NAME = 'log_results'
-    CALLBACK_NEEDS_WHITELIST = True
+def log_callback_name(func):
+    """ Add the function name to the log. """
 
-    def __init__(self):
-        super(CallbackModule, self).__init__()
```
Comment (stevekuznetsov) [inline oct/ansible/oct/callback_plugins/log_results.py]: Why did we lose this call in the new `__init__`?

Comment (bbguimaraes) [inline oct/ansible/oct/callback_plugins/log_results.py]: Completely unintentional, thanks for catching that.  Fixed.

**Comment by stevekuznetsov** [comment]
This looks really great. Lets get it in action!

**Comment by stevekuznetsov** [comment]
Thanks @bbguimaraes

**Pull request merged**